pkg/specs/upstream_test.go

```diff
@@ -0,0 +1,185 @@
+package specs
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/replicatedhq/ship/pkg/specs/githubclient"
+
+	"github.com/replicatedhq/ship/pkg/state"
+
+	"github.com/google/go-github/github"
+	"github.com/replicatedhq/ship/pkg/testing/logger"
+	"github.com/stretchr/testify/require"
+)
+
+func setupGitClient() (client *github.Client, mux *http.ServeMux, serveURL string, teardown func()) {
+	mux = http.NewServeMux()
+	server := httptest.NewServer(mux)
+	client = github.NewClient(nil)
+	url, _ := url.Parse(server.URL + "/")
+	client.BaseURL = url
+	client.UploadURL = url
+
+	return client, mux, server.URL, server.Close
+}
+
+func TestResolver_MaybeResolveVersionedUpstream(t *testing.T) {
+	tests := []struct {
+		name         string
+		upstream     string
+		currentState state.State
+		expected     string
+		expectErr    bool
+	}{
+		{
+			name:     "versioned upstream has an update available",
+			upstream: "github.com/o/r/tree/_latest_",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.1.0",
+					},
+				},
+			},
+			expected: "github.com/o/r/tree/1.2.0",
+		},
+		{
+			name:     "version upstream is above latest",
+			upstream: "github.com/o/r/tree/_latest_",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.2.1",
+					},
+				},
+			},
+			expected:  "",
+			expectErr: true,
+		},
+		{
+			name:     "not a versioned upstream",
+			upstream: "github.com/o/r/tree/ref",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.2.0",
```

Comment:
what happens with a ref upstream when no version is in the state?

Reply:
Also, should we test upstreams like "github.com/o/r/tree/ref/path" and "notgithub.com/o/r/tree/1.0.1"?

Reply:
1. ref upstream remains unchanged
2. added all the test cases